logic/concludable/ConjunctionConcludable.java

```diff
@@ -112,21 +119,107 @@ public Value asValue() {
+        }
+
+        private Stream<Map<RolePlayer, Set<RolePlayer>>> unifyRolePlayers(
+                List<RolePlayer> conjRolePlayers, List<RolePlayer> thenRolePlayers,
+                Map<RolePlayer, Set<RolePlayer>> unifier) {
+
+            int length = conjRolePlayers.size();
+            return conjRolePlayers.stream().findFirst()
+                    .flatMap(conjRP ->
+                            Optional.of(thenRolePlayers.stream()
+                                    .filter(thenRP ->
+                                            unifier.values().stream().noneMatch(rolePlayers -> rolePlayers.contains(thenRP)))
+                                    .map(thenRP -> extendRolePlayerUnifier(conjRP, thenRP, unifier))
+                                    .filter(Optional::isPresent).map(Optional::get)
+                                    .flatMap(newUnifier ->
+                                            unifyRolePlayers(conjRolePlayers.subList(1,length), thenRolePlayers, newUnifier)
+                                    ))
+            ).orElseGet(() -> Stream.of(unifier));
+
+        }
+
+        Map<Reference, Set<Reference>> convertRolePlayerUnifier(
+                Map<RolePlayer, Set<RolePlayer>> rolePlayerUnifier, Map<Reference, Set<Reference>> variableUnifier) {
+            Map<Reference, Set<Reference>> newMapping = cloneMapping(variableUnifier);
+            rolePlayerUnifier.forEach((conjRP, thenRPSet) -> {
+                thenRPSet.forEach( thenRP -> {
+                            if (conjRP.roleType().isPresent() && conjRP.roleType().get().reference().isName()) {
+                                assert thenRP.roleType().isPresent();
+                                newMapping.putIfAbsent(conjRP.roleType().get().reference(), new HashSet<>());
+                                newMapping.get(conjRP.roleType().get().reference()).add(thenRP.roleType().get().reference());
+                            }
+                            newMapping.putIfAbsent(conjRP.player().reference(), new HashSet<>());
+                            newMapping.get(conjRP.player().reference()).add(thenRP.player().reference());
+                        }
+                );
+            });
+
+            return newMapping;
+        }
+
+        private boolean roleHinstDisjoint(RolePlayer conjRP, RolePlayer thenRP) {
+            if (!conjRP.roleTypeHints().isEmpty() && !thenRP.roleTypeHints().isEmpty()
+                    && Collections.disjoint(conjRP.roleTypeHints(), thenRP.roleTypeHints())) return true;
+            if (varHintsDisjoint(conjRP.player(), thenRP.player())) return true;
+            if (conjRP.roleType().isPresent()) {
+                assert thenRP.roleType().isPresent();
+                return varHintsDisjoint(conjRP.roleType().get(), thenRP.roleType().get());
+            }
+            return false;
+        }
+
+        Optional<Map<RolePlayer, Set<RolePlayer>>> extendRolePlayerUnifier(
```

Comment:
private? can you check access modifiers everwhere? Also, public methods tend to come first and private methods later. This way you can see the class API at the very top, and the implementation follows. Also the "import" private methods i tend to put higher up as well, under the public API

Reply:
Changed to private